File: day5/main.go
package main

import (
	"flag"
	"fmt"
	"github.com/Piszmog/adventofcode/utils"
	"log"
	"sort"
	"strings"
)

type boardingPass struct {
	value  []string
	row    int
	column int
	seatId int
}

type boardingPasses []boardingPass

func (p boardingPasses) Len() int {
	return len(p)
}
func (p boardingPasses) Less(i, j int) bool {
	return p[i].seatId < p[j].seatId
}
func (p boardingPasses) Swap(i, j int) {
	p[i], p[j] = p[j], p[i]
}

func main() {
	inputFile := flag.String("f", "input.txt", "File containing the input")

	var passes boardingPasses
	err := utils.ReadTextFile(*inputFile, func(line string) error {
		passes = append(passes, boardingPass{value: strings.Split(line, "")})
		return nil
	})
	if err != nil {
		log.Fatalln(err)
	}

	// part 1
	highestSeatId := 0
	for i, pass := range passes {
		row, column := decodeValue(pass.value)
		seatId := getSeatId(row, column)
		pass.row = row
		pass.column = column
		pass.seatId = seatId
		if pass.seatId > highestSeatId {
			highestSeatId = pass.seatId
		}
		passes[i] = pass
	}
	fmt.Printf("Part 1: highest seat id: %d\n", highestSeatId)

	// part 2
	sort.Sort(passes)
	var missingSeatId int
	for i, pass := range passes {
		if i == len(passes)-1 {
			continue
		}
		// if the next seat is not +1, then +1 from current pass is the missing seat
		if pass.seatId != passes[i+1].seatId-1 {
			missingSeatId = pass.seatId + 1
		}
	}
	fmt.Printf("Part 2: Missing seat: %d\n", missingSeatId)
}

const (
	frontLowerHalf = "F"
	backUpperHalf  = "B"
	leftLowerHalf  = "L"
	rightUpperHalf = "R"

	minRow    = 0
	maxRow    = 127
	minColumn = 0
	maxColumn = 7
)

func decodeValue(value []string) (int, int) {
	currentMinRow := minRow
	currentMaxRow := maxRow
	currentMinColumn := minColumn
	currentMaxColumn := maxColumn
	for _, entry := range value {
		switch entry {
		case frontLowerHalf:
			currentMaxRow = ((currentMaxRow - currentMinRow) / 2) + currentMinRow
		case backUpperHalf:
			currentMinRow = (currentMaxRow+currentMinRow)/2 + 1
		case leftLowerHalf:
			currentMaxColumn = ((currentMaxColumn - currentMinColumn) / 2) + currentMinColumn
		case rightUpperHalf:
			currentMinColumn = (currentMaxColumn+currentMinColumn)/2 + 1
		}
	}
	return currentMinRow, currentMinColumn
}

func getSeatId(row int, column int) int {
	return row*8 + column
}

File: day5/rust/src/main.rs
fn main() {
    let file_name = "../input.txt".to_string();
    let mut input = file_utils::get_input(file_name, |line| {
        Some(BoardingPass {
            value: line.split("").map(String::from).collect(),
            row: 0,
            column: 0,
            seat_id: 0,
        })
    });

    // part 1
    let mut highest_id = 0;
    for p in &mut input {
        let value: &Vec<String> = &p.value;
        let decoded_values = decode_value(value);
        let seat_id = decoded_values.0 * 8 + decoded_values.1;
        p.seat_id = seat_id;
        p.row = decoded_values.0;
        p.column = decoded_values.1;
        if highest_id < seat_id {
            highest_id = seat_id;
        }
    }
    println!("Part 1: Highest seat id: {}", highest_id);

    // part 2
    input.sort_by(|a, b| a.seat_id.cmp(&b.seat_id));
    let mut missing_id = 0;
    for (i, p) in input.iter().enumerate() {
        if i == input.len() - 1 {
            continue;
        }
        if p.seat_id != input[i + 1].seat_id - 1 {
            missing_id = p.seat_id + 1;
            break;
        }
    }
    println!("Part 2: Missing seat id: {}", missing_id);
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
struct BoardingPass {
    value: Vec<String>,
    row: i32,
    column: i32,
    seat_id: i32,
}

fn decode_value(v: &Vec<String>) -> (i32, i32) {
    let mut current_min_row = 0;
    let mut current_max_row = 127;
    let mut current_min_column = 0;
    let mut current_max_column = 7;
    v.iter().for_each(|e| {
        match e.as_str() {
            "F" => current_max_row = (current_max_row - current_min_row) / 2 + current_min_row,
            "B" => current_min_row = (current_max_row + current_min_row) / 2 + 1,
            "L" => current_max_column = (current_max_column - current_min_column) / 2 + current_min_column,
            "R" => current_min_column = (current_max_column + current_min_column) / 2 + 1,
            _ => (),
        }
    });
    (current_max_row, current_min_column)
}